File: src/test/java/leets/weeth/domain/board/domain/repository/NoticeRepositoryTest.java
package leets.weeth.domain.board.domain.repository;

import leets.weeth.config.TestContainersConfig;
import leets.weeth.domain.board.domain.entity.Notice;
import leets.weeth.domain.board.test.fixture.NoticeTestFixture;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Import;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.data.domain.Sort;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

@DataJpaTest
@Import(TestContainersConfig.class)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
class NoticeRepositoryTest {

    @Autowired
    private NoticeRepository noticeRepository;

    @Test
    void findPageBy_공지_id_내림차순으로_조회() {
        // given
        List<Notice> notices = new ArrayList<>();
        for(int i = 0; i<5; i++){
            Notice notice = NoticeTestFixture.createNotice("공지" + i);
            notices.add(notice);
        }

        noticeRepository.saveAll(notices);
        Pageable pageable = PageRequest.of(0, 3, Sort.by(Sort.Direction.DESC, "id"));

        // when
        Slice<Notice> pagedNotices = noticeRepository.findPageBy(pageable);

        // then
        assertThat(pagedNotices.getSize()).isEqualTo(3);
        assertThat(pagedNotices)
                .extracting(Notice::getTitle)
                        .containsExactly(
                                notices.get(4).getTitle(),
                                notices.get(3).getTitle(),
                                notices.get(2).getTitle()
                        );
        assertThat(pagedNotices.hasNext()).isTrue();
    }

    @Test
    void search_검색어가_포함된_공지_id_내림차순으로_조회() {
        // given
        List<Notice> notices = new ArrayList<>();
        for(int i = 0; i<6; i++){
            Notice notice;
            if(i % 2 == 0){
                notice = NoticeTestFixture.createNotice("공지" + i);
            } else{
                notice = NoticeTestFixture.createNotice("검색" + i);
            }
            notices.add(notice);
        }

        noticeRepository.saveAll(notices);

        Pageable pageable = PageRequest.of(0, 5, Sort.by(Sort.Direction.DESC, "id"));

        // when
        Slice<Notice> searchedNotices = noticeRepository.search("검색", pageable);

        // then
        assertThat(searchedNotices.getContent()).hasSize(3);
        assertThat(searchedNotices.getContent())
                .extracting(Notice::getTitle)
                        .containsExactly(notices.get(5).getTitle(),
                                notices.get(3).getTitle(),
                                notices.get(1).getTitle());
        assertThat(searchedNotices.hasNext()).isFalse();

    }
}

File: src/test/kotlin/com/weeth/domain/board/domain/repository/NoticeRepositoryTest.kt
package com.weeth.domain.board.domain.repository

import com.weeth.config.TestContainersConfig
import com.weeth.domain.board.fixture.NoticeTestFixture
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.booleans.shouldBeFalse
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort

@DataJpaTest
@Import(TestContainersConfig::class)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
class NoticeRepositoryTest(
    private val noticeRepository: NoticeRepository,
) : DescribeSpec({

        describe("findPageBy") {
            it("공지 id 내림차순으로 조회") {
                val notices =
                    (0 until 5).map { i ->
                        NoticeTestFixture.createNotice(title = "공지$i")
                    }
                noticeRepository.saveAll(notices)

                val pageable = PageRequest.of(0, 3, Sort.by(Sort.Direction.DESC, "id"))

                val pagedNotices = noticeRepository.findPageBy(pageable)

                pagedNotices.size shouldBe 3
                pagedNotices.map { it.title } shouldContainExactly
                    listOf(notices[4].title, notices[3].title, notices[2].title)
                pagedNotices.hasNext().shouldBeTrue()
            }
        }

        describe("search") {
            it("검색어가 포함된 공지를 id 내림차순으로 조회") {
                val notices =
                    (0 until 6).map { i ->
                        if (i % 2 == 0) {
                            NoticeTestFixture.createNotice(title = "공지$i")
                        } else {
                            NoticeTestFixture.createNotice(title = "검색$i")
                        }
                    }
                noticeRepository.saveAll(notices)

                val pageable = PageRequest.of(0, 5, Sort.by(Sort.Direction.DESC, "id"))

                val searchedNotices = noticeRepository.search("검색", pageable)

                searchedNotices.content shouldHaveSize 3
                searchedNotices.content.map { it.title } shouldContainExactly
                    listOf(notices[5].title, notices[3].title, notices[1].title)
                searchedNotices.hasNext().shouldBeFalse()
            }
        }
    })
